### PR TITLE
crimson/osd: bring support for backfill cancellation

### PR DESCRIPTION
**Pull request by rzarzynski**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/osd/backfill_state.cc [hunk 1]
```diff
@@ -444,6 +465,15 @@ BackfillState::ReplicasScanning::react(ReplicaScanned evt)
   return discard_event();
 }
 
+boost::statechart::result
+BackfillState::ReplicasScanning::react(Cancel evt)
+{
+  logger().debug("{}: cancelled within ReplicasScanning",
+                 __func__);
+  waiting_on_backfill.clear();
```
Comment (xxhdx1985126) [inline src/crimson/osd/backfill_state.cc:473]: It seems that `BackfillState::progress_tracker` also need to be cleared, am I right?

File: src/crimson/osd/backfill_state.h [hunk 2]
```diff
@@ -143,13 +146,25 @@ struct BackfillState {
     sc::result react(const Triggered&);
   };
 
+  struct Cancelled : sc::state<Cancelled, BackfillMachine>,
+                     StateHelper<Cancelled> {
+    using reactions = boost::mpl::list<
+      sc::custom_reaction<Triggered>,
+      sc::transition<sc::event_base, Crashed>>;
```
Comment (xxhdx1985126) [inline src/crimson/osd/backfill_state.h:153]: I think `PrimaryScanned`, `ReplicaScanned` and `ObjectPushed` should be treated as valid events, am I right?

**Comment by athanatos** [comment]
@rzarzynski Is this still waiting for review, or did it get superseded by another PR?

**Comment by rzarzynski** [comment]
@athanatos: it awaits a review.

File: src/test/crimson/test_backfill.cc [hunk 3]
```diff
@@ -403,6 +424,68 @@ TEST(backfill, two_empty_replicas)
...
+  ).add_target(
+    { /* nothing 1 */ }
+  ).add_target(
+    { /* nothing 2 */ }
+  ).get_result();
+
+  EXPECT_CALL(cluster_fixture, backfilled);
+  cluster_fixture.next_round2<crimson::osd::BackfillState::PrimaryScanned>();
+  cluster_fixture.cancel();
+  cluster_fixture.next_round2<crimson::osd::BackfillState::Cancel>();
```
Comment (athanatos) [inline src/test/crimson/test_backfill.cc:445]: It seems like it should be possible for ReplicaScanned and ObjectPushed events to happen after Cancel as @xxhdx1985126 mentions above.

**Review by athanatos — CHANGES_REQUESTED** [review]

File: src/crimson/osd/backfill_state.h (hunk 2 above)
Comment (xxhdx1985126) [reply src/crimson/osd/backfill_state.h:153]: @rzarzynski Could you take a look at this comment?

Comment (rzarzynski) [reply src/crimson/osd/backfill_state.h:153]: @xxhdx1985126: I understood your question as _can e.g. `ReplicaScanned` event happen while the machine is in the `Cancelled`_. My very first thought is they can and they should ignored, I guess.

**Comment by Matan-B** [comment]
> Closing in favor of [#59118](https://github.com/ceph/ceph/pull/59118).

@xuxuehan, should we at least cherry-pick the tests?

File: src/crimson/osd/backfill_state.h (hunk 2 above)
Comment (xxhdx1985126) [reply src/crimson/osd/backfill_state.h:153]: > @xxhdx1985126: I understood your question as _can e.g. `ReplicaScanned` event happen while the machine is in the `Cancelled`_. My very first thought is they can and they should ignored, I guess.

Ah, yes, thanks. But this brings me to another question, what if those old events arrives when a new round of backfill has started? I also recorded this question in the comment: https://github.com/ceph/ceph/blob/44c4a176d6fc8641c00bf111e4e8f647dda36779/src/crimson/osd/pg_recovery.cc#L636-L640

**Pull request closed without merging**